src/x509_parser.rs

```diff
@@ -521,6 +551,84 @@ pub fn parse_algorithm_identifier(i: &[u8]) -> X509Result<AlgorithmIdentifier> {
     })(i)
 }
 
+fn attributes_sequence_to_map<'a>(
+    i: &'a [u8],
+    v: Vec<X509CriAttribute<'a>>,
+) -> X509Result<'a, HashMap<Oid<'a>, X509CriAttribute<'a>>> {
+    let mut attributes = HashMap::new();
+    for attr in v.into_iter() {
+        if attributes.insert(attr.oid.clone(), attr).is_some() {
+            // duplicate attributes are not allowed
+            return Err(Err::Failure(X509Error::DuplicateAttributes));
+        }
+    }
+    Ok((i, attributes))
+}
+
+fn parse_cri_attributes(i: &[u8]) -> X509Result<HashMap<Oid, X509CriAttribute>> {
+    let (i, hdr) = ber_read_element_header(i).or(Err(Err::Error(X509Error::InvalidAttributes)))?;
+    (0..hdr.structured)
+        .into_iter()
+        .try_fold((i, Vec::new()), |(i, mut attrs), _| {
+            let (rem, attr) = parse_attribute(i)?;
+            attrs.push(attr);
+            Ok((rem, attrs))
+        })
+        .and_then(|(i, attrs)| attributes_sequence_to_map(i, attrs))
+}
+
+/// Parse a certification request info structure
+///
+/// Certification request information is defined by the following ASN.1 structure:
+///
+/// <pre>
+/// CertificationRequestInfo ::= SEQUENCE {
+///      version       INTEGER { v1(0) } (v1,...),
+///      subject       Name,
+///      subjectPKInfo SubjectPublicKeyInfo{{ PKInfoAlgorithms }},
+///      attributes    [0] Attributes{{ CRIAttributes }}
+/// }
+/// </pre>
+///
+/// version is the version number; subject is the distinguished name of the certificate
+/// subject; subject_pki contains information about the public key being certified, and
+/// attributes is a collection of attributes providing additional information about the
+/// subject of the certificate.
+pub fn parse_certification_request_info<'a>(
+    i: &'a [u8],
+) -> X509Result<X509CertificationRequestInfo<'a>> {
+    let start_i = i;
+    parse_ber_sequence_defined_g(move |_, i| {
+        let (i, version) = parse_required_version(i)?;
+        let (i, subject) = parse_x509_name(i)?;
+        let (i, subject_pki) = parse_subject_public_key_info(i)?;
+        let (i, attributes) = parse_cri_attributes(i)?;
+        let len = start_i.offset(i);
+        let tbs = X509CertificationRequestInfo {
+            version,
+            subject,
+            subject_pki,
+            attributes,
+            raw: &start_i[..len],
+        };
+        Ok((i, tbs))
+    })(i)
+}
+
+pub fn parse_x509_csr_der<'a>(i: &'a [u8]) -> X509Result<X509CertificationRequest<'a>> {
```

Comment:
This function is lacking doc (and tests)

Reply:
If you focus on the other stuff, shall I focus on adding some docs and tests?

Reply:
(Or should I wait on seeing what tests you've already written?)